cmd/kubeapps-apis/plugins/fluxv2/packages/v1alpha1/chart.go

```diff
@@ -26,73 +26,179 @@ import (
+	// fields in FieldSelector is very small. things like "spec.chart" or "status.artifact.revision" are
+	// certainly not supported.
+	// see
+	//  - kubernetes/client-go#713 and
+	//  - https://github.com/flant/shell-operator/blob/8fa3c3b8cfeb1ddb37b070b7a871561fdffe788b///HOOKS.md#fieldselector and
+	//  - https://github.com/kubernetes/kubernetes/issues/53459
+	chartList, err := resourceIfc.List(ctx, metav1.ListOptions{})
 	if err != nil {
-		return false, err
-	} else if !found {
-		return false, nil
+		return nil, err
 	}
-	if generation != observedGeneration {
-		return false, nil
+	return chartList, nil
+}
+
+// returns the url from which chart .tgz can be downloaded
+// here chartVersion string, if specified at all, should be specific, like "14.4.0",
+// not an expression like ">14 <15"
+func (s *Server) getChartTarball(ctx context.Context, repoName string, chartName string, namespace string, chartVersion string) (url string, err error, cleanUp func()) {
+	// see if we the chart already exists
+	// TODO (gfichtenholt):
+	// see https://github.com/kubeapps/kubeapps/pull/2915
+	// for context. It'd be better if we could filter on server-side. The problem is the set of supported
+	// fields in FieldSelector is very small. things like "spec.chart" or "status.artifact.revision" are
+	// certainly not supported.
+	// see
+	//  - kubernetes/client-go#713 and
+	//  - https://github.com/flant/shell-operator/blob/8fa3c3b8cfeb1ddb37b070b7a871561fdffe788b///HOOKS.md#fieldselector and
+	//  - https://github.com/kubernetes/kubernetes/issues/53459
+	chartList, err := s.listChartsInCluster(ctx, namespace)
+	if err != nil {
+		return "", err, nil
 	}
 
-	conditions, found, err := unstructured.NestedSlice(unstructuredChart.Object, "status", "conditions")
+	url, err = findUrlForChartInList(chartList, repoName, chartName, chartVersion)
 	if err != nil {
-		return false, err
-	} else if !found {
-		return false, nil
-	}
-
-	// check if ready=True
-	for _, conditionUnstructured := range conditions {
-		if conditionAsMap, ok := conditionUnstructured.(map[string]interface{}); ok {
-			if typeString, ok := conditionAsMap["type"]; ok && typeString == "Ready" {
-				if statusString, ok := conditionAsMap["status"]; ok {
-					if statusString == "True" {
-						if reasonString, ok := conditionAsMap["reason"]; !ok || reasonString != "ChartPullSucceeded" {
-							// should not happen
-							log.Infof("unexpected status of HelmChart: %v", *unstructuredChart)
-						}
-						return true, nil
-					} else if statusString == "False" {
-						var msg string
-						if msg, ok = conditionAsMap["message"].(string); !ok {
-							msg = fmt.Sprintf("No message available in condition: %v", conditionAsMap)
-						}
-						// chart pull is done and it's a failure
-						return true, status.Errorf(codes.Internal, msg)
-					}
+		return "", err, nil
+	} else if url != "" {
+		return url, nil, nil
+	}
+
+	// did not find the chart, need to create
```

Comment:
I don't understand why we need to be creating a chart here at all. We're wanting to return information about the installed packages (ie. the HelmReleases). From what I've understood, if a user has created a `HelmRelease` CR, then the helm-controller would have created the related `HelmChart` for that release, so it should exist. If it doesn't, because the user has manually deleted it or whatever, then the best we can do is return the info about the installed package that we have (ie. the info from the HelmRelease).

Or what reason do you see here for a HelmRelease without the corresponding HelmChart?

Reply:
no, this is a different use case - GetAvailablePackageDetail()

Reply:
Ah - ok. I assumed this PR was just about GetInstalledPackageSummaries . Great, I'll finish going through the PR tomorrow.